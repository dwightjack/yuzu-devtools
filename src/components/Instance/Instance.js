import { wire } from 'hyperhtml';
import cc from 'classcat';
import * as styles from './Instance.styles';

const noop = () => {};

export default function Instance(props = {}) {
  const {
    Component,
    childIds,
    expanded = false,
    onClick = noop,
    onSelect = noop,
    renderChild = noop,
    selected = false,
    uid,
  } = props;
  const classes = [styles.root, { [styles.isSelected]: selected }];

  const tagStart = wire(props, ':uid')`<span
    class="${styles.tag}"
    onclick="${() => onSelect({ uid })}"
    >${Component}</span>`;
  const tagEnd = wire(props, ':uid')`<span class="${
    styles.tag
  }">${Component}</span>`;

  const expander = wire(props, ':expanded')`
    <button
      type="button"
      class="${cc([styles.expander, { [styles.isExpanderActive]: expanded }])}"
      onclick="${() => onClick({ uid, expanded: !expanded })}"
      aria-label="Expand / collapse children"
    ></button>`;

  if (Array.isArray(childIds) && childIds.length > 0) {
    classes.push({ [styles.isExpanded]: expanded });

    return wire(props)`
     <div class="${cc(classes)}">
        ${expander}${tagStart}
        <div class="${styles.childList}" hidden="${!expanded}">
          ${renderChild(childIds)}
        </div>
        ${tagEnd}
      </div>`;
  }

  return wire(props)`
    <div class="${cc(classes)}">
      ${tagStart}
    </div>`;
}
