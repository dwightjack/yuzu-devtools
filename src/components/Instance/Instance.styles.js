import { css } from 'emotion';

export const root = css`
  position: relative;
  color: var(--color-quiet);
  font-family: var(--font-monospace);
  font-size: var(--font-size-m);
  padding-left: 1rem;
  line-height: 1.5em;
  cursor: default;
  user-select: none;

  &::before {
    content: '$yuzu0';
    position: absolute;
    top: 0;
    left: calc(50% - 50vw);
    right: 0;
    z-index: -1;
    height: 1.5em;
    padding-right: 1em;
    background: var(--color-lighter);
    visibility: hidden;
    text-align: right;
    font-style: italic;
  }
`;

export const childList = css``;

export const expander = css`
  display: inline-block;
  width: 1rem;
  margin: -0.1em 0 0 -1rem;
  padding: 0;
  border: none;
  border-radius: 0;
  background: none;
  box-shadow: none;
  color: currentColor;
  font-size: 0.5rem;
  vertical-align: middle;
  text-align: center;

  &::before {
    content: '\\25b6';
  }
`;

export const tag = css`
  position: relative;
  white-space: nowrap;
  color: var(--color-accent);

  &::after,
  &::before {
    content: '';
    color: var(--color-quiet);
  }

  &::after {
    content: '\\003e';
  }
  &::before {
    content: '\\003c';
  }

  &:last-child::after {
    content: '\\002f\\003e';
  }

  & ~ &::before {
    content: '\\003c\\002f';
  }

  & ~ &::after {
    content: '\\003e';
  }
`;

export const watchMark = css`
  display: inline-block;
  width: 0.8em;
  height: 0.8em;
  margin-top: -1em;
  margin-right: 0.2em;
  vertical-align: super;
  fill: currentColor;
`;

export const isExpanderActive = css`
  &::before {
    content: '\\25bc';
  }
`;

export const isExpanded = css``;

export const isWatched = css`
  font-weight: bold;
`;

export const isSelected = css`
  &::before {
    visibility: visible;
  }
`;
