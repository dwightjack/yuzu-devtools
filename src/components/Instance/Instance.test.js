import Instance from './Instance';

describe('Instance', () => {
  test('matches default snapshot', () => {
    expect(
      Instance({
        Component: 'DemoComponent',
        uid: '_0',
      }),
    ).toMatchSnapshot();
  });

  test('matches nested snapshot', () => {
    const renderChild = () => 'CHILDREN';
    expect(
      Instance({
        Component: 'DemoComponent',
        uid: '_0',
        childIds: [0, 1],
        renderChild,
      }),
    ).toMatchSnapshot();
  });

  test('passes childIds to renderChild ', () => {
    const renderChild = jest.fn(() => '');
    const childIds = [0, 1];
    Instance({
      Component: 'DemoComponent',
      uid: '_0',
      childIds,
      renderChild,
    });
    expect(renderChild).toHaveBeenCalledWith(childIds);
  });
});
