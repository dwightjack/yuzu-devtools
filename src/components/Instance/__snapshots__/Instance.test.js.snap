// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Instance matches default snapshot 1`] = `
<div
  class="css-eus7rr"
  style=""
>
  
      
  <span
    class="css-1g7xg9f"
    style=""
  >
    DemoComponent
    
    
    
    
  </span>
  
  
    
</div>
`;

exports[`Instance matches nested snapshot 1`] = `
<div
  class="css-eus7rr"
  style=""
>
  
        
  <button
    aria-label="Expand / collapse children"
    class="css-a1ymjf"
    style=""
    type="button"
  />
  
  <span
    class="css-1g7xg9f"
    style=""
  >
    DemoComponent
    
    
    
    
  </span>
  
  
        
  <div
    class="css-0"
    hidden={true}
    style=""
  >
    
          
    CHILDREN
    
    
        
  </div>
  
        
  <span
    class="css-1g7xg9f"
    style=""
  >
    DemoComponent
    
  </span>
  
  
      
</div>
`;

exports[`Instance passes childIds to renderChild  1`] = `
<div
  class="css-eus7rr"
  style=""
>
  
        
  <button
    aria-label="Expand / collapse children"
    class="css-a1ymjf"
    style=""
    type="button"
  />
  
  <span
    class="css-1g7xg9f"
    style=""
  >
    DemoComponent
    
    
    
    
  </span>
  
  
        
  <div
    class="css-0"
    hidden={true}
    style=""
  >
    
          
    CHILDREN
    
    
        
  </div>
  
        
  <span
    class="css-1g7xg9f"
    style=""
  >
    DemoComponent
    
  </span>
  
  
      
</div>
`;
